clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3431,23 +3416,96 @@ static void handleSYCLIntelSchedulerTargetFmaxMhzAttr(Sema &S, Decl *D,
 }
 
 // Handles max_global_work_dim.
-static void handleMaxGlobalWorkDimAttr(Sema &S, Decl *D, const ParsedAttr &A) {
-  if (D->isInvalidDecl())
-    return;
-
-  Expr *E = A.getArgAsExpr(0);
+void Sema::AddSYCLIntelMaxGlobalWorkDimAttr(Decl *D,
+                                            const AttributeCommonInfo &CI,
+                                            Expr *E) {
+  if (!E->isValueDependent()) {
+    // Validate that we have an integer constant expression and then store the
+    // converted constant expression into the semantic attribute so that we
+    // don't have to evaluate it again later.
+    llvm::APSInt ArgVal;
+    ExprResult Res = VerifyIntegerConstantExpression(E, &ArgVal);
+    if (Res.isInvalid())
+      return;
+    E = Res.get();
 
-  if (!checkWorkGroupSizeValues(S, D, A)) {
-    D->setInvalidDecl();
-    return;
+    // This attribute requires a non-negative value.
+    if (ArgVal < 0) {
+      Diag(E->getExprLoc(), diag::err_attribute_requires_positive_integer)
+          << CI << /*non-negative*/ 1;
+      return;
+    }
+    // This attribute must be in the range [0, 3].
+    if (ArgVal > 3) {
+      Diag(E->getBeginLoc(), diag::err_attribute_argument_out_of_range)
+          << CI << 0 << 3 << E->getSourceRange();
+      return;
+    }
+    // Check to see if there's a duplicate attribute with different values
+    // already applied to the declaration.
+    if (const auto *DeclAttr = D->getAttr<SYCLIntelMaxGlobalWorkDimAttr>()) {
+      // If the other attribute argument is instantiation dependent, we won't
+      // have converted it to a constant expression yet and thus we test
+      // whether this is a null pointer.
+      const auto *DeclExpr = dyn_cast<ConstantExpr>(DeclAttr->getValue());
+      if (DeclExpr && ArgVal != DeclExpr->getResultAsAPSInt()) {
+        Diag(CI.getLoc(), diag::warn_duplicate_attribute) << CI;
+        Diag(DeclAttr->getLoc(), diag::note_previous_attribute);
+        return;
```

Comment:
@AaronBallman, I created #3336 . I plan to address this duplicate attributes issues in separate PR  since other attributes also have same issues like https://github.com/intel/llvm/pull/3224. Does it sound good to you?

Reply:
Because you're already refactoring this code to fix diagnostic issues with this attribute, I think it probably makes more sense (to me, at least) to make the changes now so that it's one less task for the future. However, I don't have strong opinions about it, so I leave it to @elizabethandrews and @premanandrao if they have a preference.

Reply:
>>Because you're already refactoring this code to fix diagnostic issues with this attribute, I think it probably makes more sense (to me, at least) to make the changes now so that it's one less task for the future. 

Thanks for the input. Yes, this sounds reasonable to me. I will address it and upload a new patch here.

Reply:
Done

Reply:
> We don't want to duplicate the attribute with the same arguments, so I think the code should be more like:
> 
> ```
> if (const auto *DeclExpr = dyn_cast<ConstantExpr>(DeclAttr->getValue())) {
>   if (ArgVal != DeclExpr->getResultAsAPSInt()) {
>     ...
>   }
>   return;
> }
> ```

With this change,  we will still duplicate attribute with the same arguments.  i am looking into it.

Reply:
> We don't want to duplicate the attribute with the same arguments, so I think the code should be more like:
> 
> ```
> if (const auto *DeclExpr = dyn_cast<ConstantExpr>(DeclAttr->getValue())) {
>   if (ArgVal != DeclExpr->getResultAsAPSInt()) {
>     ...
>   }
>   return;
> }
> ```

@AaronBallman, I have added D->dropAttr<>(); to drop duplicate attribute when adding and merging since only "return" statement does not drop the duplicate attribute here.

Reply:
Done